File: LeetCode/Java/589-Unknown-n叉树的前序遍历.java
import java.util.List;
import java.util.ArrayList;
import java.util.Queue;

/*
 * @lc app=leetcode.cn id=589 lang=java
 *
 * [589] N叉树的前序遍历
 */

// @lc code=start
class Solution {
    private List<Integer> res = new ArrayList();

    public List<Integer> preorder(Node root) {
        _preorder(root);
        return res;
    }

    private void _preorder(Node root) {
        if (root == null)
            return;
        res.add(root.val);
        for (Node n : root.children)
            _preorder(n);
    }
}
// @lc code=end

File: LeetCode/C/589-Unknown-n叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=589 lang=cpp
 *
 * [589] N叉树的前序遍历
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    vector<int> preorder(Node *root) {
        if (root == nullptr)
            return {};
        vector<int> res;
        stack<Node *> st;
        st.push(root);
        while (!st.empty()) {
            Node *t = st.top();
            st.pop();
            res.push_back(t->val);
            for (int i = t->children.size() - 1; i >= 0; --i)
                if (t->children[i])
                    st.push(t->children[i]);
        }
        return res;
    }
};
// @lc code=end
